File: bisseccao.java
import java.util.Scanner;
import java.util.function.DoubleUnaryOperator;

public class MetodoBissecaoPolinomio {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("================================================");
        System.out.println("MÉTODO DA BISSECÇÃO - ZEROS DE FUNÇÕES POLINOMIAIS");
        System.out.println("================================================\n");

        System.out.println("Digite o grau do polinômio (0 a 10):");
        int grau = scanner.nextInt();
        
        if (grau < 0 || grau > 10) {
            System.out.println("Erro: Grau deve ser entre 0 e 10.");
            scanner.close();
            return;
        }
        
        double[] coeficientes = new double[grau + 1];
        
        System.out.println("\nDigite os coeficientes do termo de maior para menor grau:");
        for (int i = grau; i >= 0; i--) {
            System.out.printf("Coeficiente de x^%d: ", i);
            coeficientes[i] = scanner.nextDouble();
        }

        System.out.println("\nPolinômio inserido:");
        System.out.println(formatarPolinomio(coeficientes));

        double erro;
        do {
            System.out.println("\nDigite a precisão desejada (ex: 0.0001):");
            erro = scanner.nextDouble();
            
            if (erro <= 0) {
                System.out.println("Erro: A precisão deve ser maior que zero!");
            }
        } while (erro <= 0);

        double inicio, fim;
        do {
            System.out.println("\nDigite o limite inferior do intervalo:");
            inicio = scanner.nextDouble();
            System.out.println("Digite o limite superior do intervalo:");
            fim = scanner.nextDouble();
            
            if (inicio >= fim) {
                System.out.println("Erro: O limite inferior deve ser menor que o superior!");
            }
        } while (inicio >= fim);

        DoubleUnaryOperator funcao = x -> {
            double resultado = 0;
            for (int i = 0; i <= grau; i++) {
                resultado += coeficientes[i] * Math.pow(x, i);
            }
            return resultado;
        };

        System.out.println("\n================================================");
        System.out.printf("RESULTADOS (precisão < %.2e)\n", erro);
        System.out.println("================================================");

        encontrarRaizes(funcao, inicio, fim, erro);
        
        scanner.close();
    }

    private static String formatarPolinomio(double[] coeficientes) {
        StringBuilder sb = new StringBuilder();
        boolean primeiroTermo = true;
        
        for (int i = coeficientes.length - 1; i >= 0; i--) {
            if (coeficientes[i] != 0) {
                if (!primeiroTermo) {
                    sb.append(coeficientes[i] > 0 ? " + " : " - ");
                } else if (coeficientes[i] < 0) {
                    sb.append("-");
                }
                
                sb.append(Math.abs(coeficientes[i]));
                if (i > 0) {
                    sb.append("x");
                    if (i > 1) {
                        sb.append("^").append(i);
                    }
                }
                primeiroTermo = false;
            }
        }
        return sb.toString();
    }

    public static void encontrarRaizes(DoubleUnaryOperator funcao, double inicio, double fim, double erro) {
        final int DIVISOES = 50; 
        double passo = (fim - inicio) / DIVISOES;
        boolean raizEncontrada = false;
        
        System.out.printf("\nProcurando raízes em [%.4f, %.4f]...\n", inicio, fim);

        for (double a = inicio; a < fim; a += passo) {
            double b = Math.min(a + passo, fim);
            double fa = funcao.applyAsDouble(a);
            double fb = funcao.applyAsDouble(b);
            
            if (Double.isNaN(fa) || Double.isNaN(fb)) {
                System.out.printf("Aviso: Função indefinida em [%.4f, %.4f]\n", a, b);
                continue;
            }

            if (fa * fb <= 0) {
                System.out.printf("\n● Raiz detectada em [%.6f, %.6f]\n", a, b);
                raizEncontrada = true;
                
                double raiz = bissecao(funcao, a, b, erro);
                double valorNaRaiz = funcao.applyAsDouble(raiz);
                
                System.out.printf("\n✔ Raiz aproximada: %.10f\n", raiz);
                System.out.printf("  f(raiz) = %.5e (precisão: %.5e)\n", 
                                valorNaRaiz, Math.abs(valorNaRaiz));
                System.out.println("----------------------------------------");
            }
        }
        
        if (!raizEncontrada) {
            System.out.println("\nNenhuma raiz encontrada no intervalo especificado.");
            System.out.println("Sugestões:");
            System.out.println("- Amplie o intervalo de busca");
            System.out.println("- Verifique se o polinômio possui raízes reais");
        }
    }

    public static double bissecao(DoubleUnaryOperator funcao, double a, double b, double erro) {
        final int MAX_ITERACOES = 1000;
        double fa = funcao.applyAsDouble(a);
        double fb = funcao.applyAsDouble(b);
        
        if (fa * fb > 0) {
            throw new IllegalArgumentException("A função deve ter sinais opostos nos extremos do intervalo");
        }
        
        int iteracoes = 0;
        double x = 0;
        double erroAtual = Double.POSITIVE_INFINITY;
        
        System.out.println("\nIteração |   a         |   b         |   x         |   f(x)         | Erro Atual");
        System.out.println("-----------------------------------------------------------------------------");
        
        do {
            double xAnterior = x;
            x = (a + b) / 2;
            double fx = funcao.applyAsDouble(x);
            
            erroAtual = Math.abs(b - a);
            
            System.out.printf("%5d    | %10.8f | %10.8f | %10.8f | %14.8e | %10.8e\n", 
                            iteracoes, a, b, x, fx, erroAtual);
            
            if (fx == 0) {
                return x;
            } else if (fa * fx < 0) {
                b = x;
                fb = fx;
            } else {
                a = x;
                fa = fx;
            }
            
            iteracoes++;

            if (iteracoes >= MAX_ITERACOES) {
                System.out.println("\nAviso: Máximo de iterações atingido!");
                break;
            }
            
        } while (erroAtual > erro);
        
        return x;
    }
}

File: projeto-matematica/js/bisseccao.js
document.addEventListener('DOMContentLoaded', function() {
    const grauInput = document.getElementById('grau');
    const coeficientesContainer = document.getElementById('coeficientes-container');
    const calcularBtn = document.getElementById('calcular-btn');
    const polinomioDiv = document.getElementById('polinomio');
    const resultadosDiv = document.getElementById('resultados');

    grauInput.addEventListener('change', atualizarCoeficientes);
    atualizarCoeficientes();
    calcularBtn.addEventListener('click', calcularBisseccao);

    function atualizarCoeficientes() {
        const grau = parseInt(grauInput.value);
        coeficientesContainer.innerHTML = '';

        for (let i = grau; i >= 0; i--) {
            const div = document.createElement('div');
            div.className = 'mb-2';

            const label = document.createElement('label');
            label.className = 'form-label';
            label.textContent = `Coeficiente de x^${i}:`;

            const input = document.createElement('input');
            input.type = 'number';
            input.className = 'form-control';
            input.step = 'any';
            input.value = i === grau ? '1' : '0';
            input.id = `coef-${i}`;

            div.appendChild(label);
            div.appendChild(input);
            coeficientesContainer.appendChild(div);
        }
    }

    function formatarPolinomio(coeficientes) {
        let resultado = '';
        let primeiroTermo = true;

        for (let i = coeficientes.length - 1; i >= 0; i--) {
            if (coeficientes[i] !== 0) {
                if (!primeiroTermo) {
                    resultado += coeficientes[i] > 0 ? ' + ' : ' - ';
                } else if (coeficientes[i] < 0) {
                    resultado += '-';
                }

                const absCoef = Math.abs(coeficientes[i]);
                resultado += absCoef === 1 && i > 0 ? '' : absCoef;

                if (i > 0) {
                    resultado += 'x';
                    if (i > 1) {
                        resultado += '^' + i;
                    }
                }
                primeiroTermo = false;
            }
        }

        return resultado || '0';
    }

    function calcularBisseccao() {
        const grau = parseInt(grauInput.value);
        const erro = parseFloat(document.getElementById('erro').value);
        const inicio = -1000;
        const fim = 1000;

        const coeficientes = [];
        for (let i = 0; i <= grau; i++) {
            coeficientes[i] = parseFloat(document.getElementById(`coef-${i}`).value) || 0;
        }

        polinomioDiv.innerHTML = `<strong>Polinômio:</strong> P(x) = ${formatarPolinomio(coeficientes)}`;

        const funcao = x => {
            let resultado = 0;
            for (let i = 0; i <= grau; i++) {
                resultado += coeficientes[i] * Math.pow(x, i);
            }
            return resultado;
        };

        const DIVISOES = 50;
        const passo = (fim - inicio) / DIVISOES;
        let output = '<h3>Passo a Passo do Método da Bissecção</h3><pre>';
        let raizEncontrada = false;

        for (let a = inicio; a < fim; a += passo) {
            let b = Math.min(a + passo, fim);
            let fa = funcao(a);
            let fb = funcao(b);

            if (fa * fb <= 0) {
                raizEncontrada = true;
                output += `\n--- Raiz detectada no intervalo [${a.toFixed(6)}, ${b.toFixed(6)}] ---\n`;
                const resultadoBisseccao = bissecaoComPassos(funcao, a, b, erro);
                output += resultadoBisseccao.log;
                output += `\n✔ Raiz aproximada: ${resultadoBisseccao.raiz.toFixed(10)}\n`;
                output += `f(raiz) = ${funcao(resultadoBisseccao.raiz).toExponential(5)}\n`;
            }
        }

        if (!raizEncontrada) {
            output += "\nNenhuma raiz encontrada no intervalo especificado.\n";
            output += "Sugestões:\n";
            output += "- Amplie o intervalo de busca\n";
            output += "- Verifique se o polinômio possui raízes reais\n";
        }

        output += '</pre>';
        resultadosDiv.innerHTML = output;
    }

    function bissecaoComPassos(funcao, a, b, erro) {
        const MAX_ITERACOES = 1000;
        let log = '';
        let iteracoes = 0;
        let x = 0;
        let erroAtual = Number.POSITIVE_INFINITY;

        log += `Intervalo inicial: [${a.toFixed(6)}, ${b.toFixed(6)}]\n`;

        do {
            const fa = funcao(a);
            const fb = funcao(b);

            if (fa * fb > 0) {
                log += `Aviso: A função não tem sinais opostos nos extremos de [${a.toFixed(6)}, ${b.toFixed(6)}].\n`;
                return { raiz: NaN, log: log };
            }

            x = (a + b) / 2;
            const fx = funcao(x);
            erroAtual = Math.abs(b - a);
            iteracoes++;

            log += `\nIteração ${iteracoes}:\n`;
            log += `  a = ${a.toFixed(10)}, b = ${b.toFixed(10)}\n`;
            log += `  Ponto médio c = (a + b) / 2 = ${x.toFixed(10)}\n`;
            log += `  f(a) = ${fa.toExponential(5)}, f(c) = ${fx.toExponential(5)}\n`;
            
            if (fx === 0) {
                log += `  f(c) é zero, raiz exata encontrada.\n`;
                break;
            }

            if (fa * fx < 0) {
                b = x;
                log += `  f(a) e f(c) têm sinais opostos. Novo intervalo: [a, c] -> [${a.toFixed(10)}, ${b.toFixed(10)}]\n`;
            } else {
                a = x;
                log += `  f(c) e f(b) têm sinais opostos. Novo intervalo: [c, b] -> [${a.toFixed(10)}, ${b.toFixed(10)}]\n`;
            }

            log += `  Erro atual: |b - a| = ${erroAtual.toExponential(5)}\n`;

            if (iteracoes >= MAX_ITERACOES) {
                log += `\nAviso: Máximo de iterações atingido antes da precisão desejada.\n`;
                break;
            }
        } while (erroAtual > erro);

        return { raiz: x, log: log };
    }
});
